File: Problem2.py
'''

Time Complexity : O(mn)
Space Complexity : O(mn)
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No issues faced

Your code here along with comments explaining your approach:
Using BFS here. I start by adding all the indexes of 0's to the queue and then go on visiting all the unvisited nodes


'''

class Solution:
    def updateMatrix(self, image: List[List[int]]) -> List[List[int]]:

        dir = [(0,1),(1,0),(0,-1),(-1,0)]
        m = len(image)
        n = len(image[0])
        q = collections.deque()

        for i in range(len(image)):
            for j in range(len(image[0])):

                if image[i][j] == 1:
                    image[i][j] = -1

                else:
                    q.append(i)
                    q.append(j)

        while q:
            cr = q.popleft()
            cc = q.popleft()
            for k in dir:
                ni = cr + k[0]
                nj = cc + k[1]
                if 0 <= ni < m and 0 <= nj < n and image[ni][nj] == -1:
                    q.append(ni)
                    q.append(nj)
                    image[ni][nj] = image[cr][cc] + 1

        return image

File: Problem2.cs
public class NearestZero
    {
        // Time Complexity : O(m * n) - have to traverse entire matrix
        // Space Complexity : O(m * n) - queue will have almost all elements from matrix in the worst case 
        // Did this code successfully run on Leetcode : Yes
        // Any problem you faced while coding this : No
        int[][] directions;
        public int[][] UpdateMatrix(int[][] matrix)
        {
            if (matrix == null || matrix.Length == 0) return matrix;

            int m = matrix.Length;
            int n = matrix[0].Length;

            Queue<int[]> q = new Queue<int[]>();
            directions = new int[][] {
                new int[] { 0, 1 }, //right
                new int[] { 0, -1 }, //left
                new int[] { 1, 0 }, //down
                new int[] { -1, 0 } //up
            };

            //adding all 0s into the queue and 1s to -1
            for (int i = 0; i < m; i++)
            {
                for(int j = 0; j < n; j++)
                {
                    if (matrix[i][j] == 0)
                        q.Enqueue(new int[] { i, j });
                    else
                        matrix[i][j] = -1;
                }
            }
            int distance = 1;
            while(q.Count > 0)
            {
                int size = q.Count;
                for(int i = 0; i < size; i++)
                {
                    int[] curr = q.Dequeue();
                    foreach (var dir in directions)
                    {
                        int nr = curr[0] + dir[0];
                        int nc = curr[1] + dir[1];

                        //bounds check
                        if(nr >= 0 && nr < m && nc >=0 && nc < n && matrix[nr][nc] == -1)
                        {
                            q.Enqueue(new int[] { nr, nc });
                            matrix[nr][nc] = distance;
                        }
                    }
                }
                //increment the distance once level is complete
                distance++;
            }



            return matrix;
        }
}
